api/webhook/validate_auth.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2021 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webhook
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"reflect"
+
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	authv1 "k8s.io/api/authorization/v1"
+	authorizationv1 "k8s.io/client-go/kubernetes/typed/authorization/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
+	"github.com/chaos-mesh/chaos-mesh/pkg/selector"
+)
+
+var authLog = ctrl.Log.WithName("validate-auth")
+
+// +kubebuilder:webhook:path=/validate-auth,mutating=false,failurePolicy=fail,groups=chaos-mesh.org,resources=*,verbs=create;update,versions=v1alpha1,name=vauth.kb.io
+
+// AuthValidator validates the authority
+type AuthValidator struct {
+	Enable  bool
+	Client  client.Client
+	Reader  client.Reader
+	AuthCli *authorizationv1.AuthorizationV1Client
+
+	decoder *admission.Decoder
+
+	ClusterScoped     bool
+	TargetNamespace   string
+	AllowedNamespaces string
+	IgnoredNamespaces string
+}
+
+// AuthValidator admits a pod iff a specific annotation exists.
+func (v *AuthValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	if !v.Enable {
+		return admission.Allowed("")
+	}
+
+	needAuth := true
+	username := req.UserInfo.Username
+	groups := req.UserInfo.Groups
+	chaosKind := req.Kind.Kind
+
+	var chaos, oldChaos v1alpha1.ChaosValidator
+
+	switch chaosKind {
```

Comment:
That might not be a very good pratice to write such code every time we want to get chaos type from type string. Don't we have some existing function for that?

Reply:
add a new function `GetChaosValidator`